x/ibc/02-client/keeper/client.go

```diff
@@ -71,9 +71,15 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, header exported.H
 
 	switch clientType {
 	case exported.Tendermint:
-		trustedConsState, found := k.GetClientConsensusStateLTE(ctx, clientID, header.GetHeight())
+		tmHeader, ok := header.(ibctmtypes.Header)
+		if !ok {
+			err = sdkerrors.Wrapf(types.ErrInvalidHeader, "expected tendermint header: %T, got header type: %T", ibctmtypes.Header{}, header)
+			break
+		}
+		// Get the consensus state at the trusted height of header
+		trustedConsState, found := k.GetClientConsensusState(ctx, clientID, tmHeader.TrustedHeight)
```

Comment:
For follow-up issue/PR. Since the solo machine client also contains a [`Header`](https://github.com/cosmos/ics/tree/master/spec/ics-006-solo-machine-client#headers) type, I wonder if we can add it to the Header interface an avoid the casting above?

Reply:
we should just pass in the client's store (like we do in the verify funcs) into `CheckValidityAndUpdateState` which would be an interface func in client when we modularize it. Ditto for misbehavior. This would allow the client to retrieve whatever consensus states they need to prove an update or misbehaviour